File: app/src/main/java/com/castleglobal_clean/utils/SearchResultsHelper.java
package com.castleglobal_clean.utils;

import com.castleglobal_clean.domain.model.Restaurant;
import com.castleglobal_clean.domain.model.RestaurantWrapper;
import com.castleglobal_clean.domain.model.SearchResult;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

/**
 * Created by sahil on 3/10/18.
 */

public class SearchResultsHelper {

    public static HashMap<String, ArrayList<Restaurant>> getRestaurantsByCuisineType(List<RestaurantWrapper> restaurantsList) {
        HashMap<String, ArrayList<Restaurant>> restaurantsByCuisine = new HashMap<String, ArrayList<Restaurant>>();

        for (RestaurantWrapper restaurantWrapper : restaurantsList) {
            Restaurant restaurant = restaurantWrapper.getRestaurant();
            String cuisinesOffered = restaurant.getCuisines();
            String[] cuisinesList = cuisinesOffered.split(", ");
            for (String cuisineOffered : cuisinesList) {
                ArrayList<Restaurant> cuisineOfferingRestaurant;
                if (restaurantsByCuisine.containsKey(cuisineOffered)) {
                    cuisineOfferingRestaurant = restaurantsByCuisine.get(cuisineOffered);
                    cuisineOfferingRestaurant.add(restaurant);
                } else {
                    cuisineOfferingRestaurant = new ArrayList<Restaurant>();
                    cuisineOfferingRestaurant.add(restaurant);
                }
                restaurantsByCuisine.put(cuisineOffered, cuisineOfferingRestaurant);
            }
        }
        return restaurantsByCuisine;
    }

    public static ArrayList<SearchResult> getSearchResultList(HashMap<String, ArrayList<Restaurant>> restaurantsByCuisines) {
        ArrayList<SearchResult> searchResultsList = new ArrayList<>();
        ArrayList<String> cuisinesList = new ArrayList<>(restaurantsByCuisines.keySet());
        for (String cuisine : cuisinesList) {
            SearchResult searchResult = new SearchResult();
            searchResult.setRestaurant(false);
            searchResult.setName(cuisine);
            searchResultsList.add(searchResult);
            ArrayList<Restaurant> restaurants = restaurantsByCuisines.get(cuisine);
            for (Restaurant restaurant : restaurants) {
                searchResult = new SearchResult();
                searchResult.setRestaurant(true);
                searchResult.setRestaurant(restaurant);
                searchResultsList.add(searchResult);
            }
        }
        return searchResultsList;
    }
}

File: domain/src/main/java/com/demo/domain/search/utils/SearchResultsHelper.kt
package com.demo.domain.search.utils

import com.demo.domain.search.model.Restaurant
import com.demo.domain.search.model.RestaurantWrapper
import com.demo.domain.search.model.SearchResult
import java.util.*
import javax.inject.Inject

class SearchResultsHelper @Inject constructor(){

    fun getRestaurantsByCuisineType(restaurantsList: List<RestaurantWrapper>): Map<String, List<Restaurant>> {
        val restaurantsByCuisine = HashMap<String, ArrayList<Restaurant>>()

        for ((restaurant) in restaurantsList) {
            val cuisinesOffered = restaurant.cuisines
            val cuisinesList = cuisinesOffered.split(", ".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
            for (cuisineOffered in cuisinesList) {
                val cuisineOfferingRestaurants: ArrayList<Restaurant>?
                if (restaurantsByCuisine.containsKey(cuisineOffered)) {
                    cuisineOfferingRestaurants = restaurantsByCuisine[cuisineOffered]
                    cuisineOfferingRestaurants!!.add(restaurant)
                } else {
                    cuisineOfferingRestaurants = ArrayList()
                    cuisineOfferingRestaurants.add(restaurant)
                }
                restaurantsByCuisine[cuisineOffered] = cuisineOfferingRestaurants
            }
        }
        return restaurantsByCuisine
    }

    fun getSearchResultList(restaurantsByCuisines: Map<String, List<Restaurant>>): List<SearchResult>{
        val searchResultsList = ArrayList<SearchResult>()
        val cuisinesList = ArrayList(restaurantsByCuisines.keys)
        for (cuisine in cuisinesList) {
            var searchResult = SearchResult(false, Optional.of(cuisine), Optional.empty())
            searchResultsList.add(searchResult)
            val restaurants = restaurantsByCuisines[cuisine]
            for (restaurant in restaurants!!) {
                searchResult = SearchResult(true, Optional.empty(), Optional.of(restaurant))
                searchResultsList.add(searchResult)
            }
        }
        return searchResultsList
    }

}